### ISSUE TITLE
Common interface for all metrics

### ISSUE THREAD

**Issue by jack89roberts** [alan-turing-institute/ARC-LoCoMoSeT#44]
The task-agnostic metrics (e.g. #10) will break the assumption that metrics take image features and labels as input. We probably want all metrics to work in the form `metric(dataset, model) -> value` instead. This could include implementing a `Metric` class that all metrics inherit from.

**Comment by eddableheath** [comment]
I think that the results should all include the task-agnostic metric in them as standard. This will allow that to be used in #37 more easily as this can be done as post-processing of the results.

The common interface is still necessary as #35 is very different to the others in the way it functions.

**Comment by eddableheath** [comment]
For the time being I've adapted the run.py script to apply the function differently (and not load any of the data or processor etc.) in  https://github.com/alan-turing-institute/ARC-LoCoMoSeT/commit/a91f9f1dc0132569a11db2927b1e19d0e74ea494

**Comment by jack89roberts** [comment]
Documenting the inputs of the various metrics we've considered so far:

- Model only:
  - No. parameters
- Model features on new dataset & ground truth labels of new dataset:
  - Renggli
  - PARC
  - LogME
  - H-Score
- Model distribution over source dataset label set (of new datapoints) & ground truth labels of new dataset:
  - LEEP
- Model labels on pre-training dataset & ground truth labels of new dataset:
  - NCE

**Comment by eddableheath** [comment]
I'm going to write up a pull request for the alternative metrics, but I think it's not worth me rewriting `run.py` for the sake of NCE as that will be covered by this.

**Comment by eddableheath** [comment]
Tie #51 into this as an efficiency consideration.

**Comment by eddableheath** [comment]
![model_experiment_structure](https://github.com/alan-turing-institute/ARC-LoCoMoSeT/assets/8882079/52808eb2-3f73-4399-9d0f-aac6b202bff3)

Rough outline of proposed structure, with the key feature being that each single 'experiment' is now per model, with all the metrics being compared noted in the config. This more easily allows for sharing inference data across multiple metrics. This will be instantiated and called from an overall run script, with the idea that if we wish to compare multiple models the run script will create a model experiment for each one.

**Comment by eddableheath** [comment]
NCE is a special case in that it requires the source dataset, before implementing this into the above design we should discuss whether or not we're going to continue with it as a metric since the source dataset might not always be available.

**Comment by eddableheath** [comment]
Todo:

- [ ] Decide on use of LEEP and NCE.
- [ ] Either adapt implementation for this or remove notes from docstrings.

**Comment by eddableheath** [comment]
WIP pipeline is now functional, see the branch for the new config file type example `configs/config_pets_model_expr_test.ymal`. Further design choices to discuss:

- [ ] Inclusion of LEEP and NCE.
- [ ] Parameter sweep options.

Todo:

- [ ] Tests.
- [ ] Tidy up/remove old implementations.
